subsys/bluetooth/host/hci_core.c

```diff
@@ -3819,81 +4508,331 @@ void bt_data_parse(struct net_buf_simple *ad,
+{
+	struct bt_le_scan_cb *listener;
+	struct net_buf_simple_state state;
+	bt_addr_le_t id_addr;
+
+	BT_DBG("%s event %u, len %u, rssi %d dBm", bt_addr_le_str(addr),
+	       info->adv_type, len, info->rssi);
+
+	if (!IS_ENABLED(CONFIG_BT_PRIVACY) &&
+	    !IS_ENABLED(CONFIG_BT_SCAN_WITH_IDENTITY) &&
+	    atomic_test_bit(bt_dev.flags, BT_DEV_EXPLICIT_SCAN) &&
+	    (info->adv_props & BT_HCI_LE_ADV_PROP_DIRECT)) {
+		BT_DBG("Dropped direct adv report");
+		return;
+	}
+
+	if (addr->type == BT_ADDR_LE_PUBLIC_ID ||
+	    addr->type == BT_ADDR_LE_RANDOM_ID) {
+		bt_addr_le_copy(&id_addr, addr);
+		id_addr.type -= BT_ADDR_LE_PUBLIC_ID;
+	} else if (addr->type == BT_HCI_PEER_ADDR_ANONYMOUS) {
+		bt_addr_le_copy(&id_addr, BT_ADDR_LE_ANY);
+	} else {
+		bt_addr_le_copy(&id_addr,
+				bt_lookup_id_addr(BT_ID_DEFAULT, addr));
+	}
+
+	info->addr = &id_addr;
+
+	if (scan_dev_found_cb) {
+		net_buf_simple_save(&buf->b, &state);
+
+		buf->len = len;
+		scan_dev_found_cb(&id_addr, info->rssi, info->adv_type,
+				  &buf->b);
+
+		net_buf_simple_restore(&buf->b, &state);
+	}
+
+
+	SYS_SLIST_FOR_EACH_CONTAINER(&scan_cbs, listener, node) {
+		net_buf_simple_save(&buf->b, &state);
+
+		buf->len = len;
+		listener->recv(info, &buf->b);
+
+		net_buf_simple_restore(&buf->b, &state);
+	}
+
+#if defined(CONFIG_BT_CENTRAL)
+	check_pending_conn(&id_addr, addr, info->adv_props);
+#endif /* CONFIG_BT_CENTRAL */
+}
+
+#if defined(CONFIG_BT_EXT_ADV)
+static void le_scan_timeout(struct net_buf *buf)
+{
+	struct bt_le_scan_cb *listener;
+
+	atomic_clear_bit(bt_dev.flags, BT_DEV_SCANNING);
+	atomic_clear_bit(bt_dev.flags, BT_DEV_EXPLICIT_SCAN);
+
+	atomic_clear_bit(bt_dev.flags, BT_DEV_SCAN_LIMITED);
+	atomic_clear_bit(bt_dev.flags, BT_DEV_RPA_VALID);
+
+#if defined(CONFIG_BT_SMP)
+	pending_id_keys_update();
+#endif
+
+	SYS_SLIST_FOR_EACH_CONTAINER(&scan_cbs, listener, node) {
+		listener->timeout();
+	}
+}
+
+/* Convert Extended adv report evt_type field into adv type */
+static u8_t get_adv_type(u8_t evt_type)
+{
+	switch (evt_type) {
+	case BT_HCI_LE_ADV_PROP_CONN |
+	     BT_HCI_LE_ADV_PROP_SCAN |
+	     BT_HCI_LE_ADV_PROP_LEGACY:
+		return BT_GAP_ADV_TYPE_ADV_IND;
+
+	case BT_HCI_LE_ADV_PROP_CONN |
+	     BT_HCI_LE_ADV_PROP_DIRECT |
+	     BT_HCI_LE_ADV_PROP_LEGACY:
+		return BT_GAP_ADV_TYPE_ADV_DIRECT_IND;
+
+	case BT_HCI_LE_ADV_PROP_SCAN |
+	     BT_HCI_LE_ADV_PROP_LEGACY:
+		return BT_GAP_ADV_TYPE_ADV_SCAN_IND;
+
+	case BT_HCI_LE_ADV_PROP_LEGACY:
+		return BT_GAP_ADV_TYPE_ADV_NONCONN_IND;
+
+	case BT_HCI_LE_ADV_PROP_SCAN_RSP |
+	     BT_HCI_LE_ADV_PROP_CONN |
+	     BT_HCI_LE_ADV_PROP_SCAN |
+	     BT_HCI_LE_ADV_PROP_LEGACY:
+	case BT_HCI_LE_ADV_PROP_SCAN_RSP |
+	     BT_HCI_LE_ADV_PROP_SCAN |
+	     BT_HCI_LE_ADV_PROP_LEGACY:
+		/* Scan response from connectable or non-connectable advertiser.
+		 */
+		return BT_GAP_ADV_TYPE_SCAN_RSP;
+
+	default:
+		return BT_GAP_ADV_TYPE_EXT_ADV;
+	}
+}
+
+static u8_t get_phy(u8_t hci_phy)
+{
+	return hci_phy ? BIT(hci_phy) : 0;
```

Comment:
I think I understand what you're trying to do here (and why), but it has a few issues:

 - People looking at the PHY definitions in bluetooth.h will wonder why it starts from `BIT(1)` rather than `BIT(0)`
 - If you end up with a newer controller (supporting some future core spec version) or the controller gives you garbage (a value >= 0x03), you'll end up giving garbage to the application
 - The readability of the code hurts, i.e. some additional research is needed by the reader to understand the conversion and to verify whether it is correct.

I don't think these sacrifices are justifiable for the benefit of slightly more efficient code, I.e. I'd suggest you to do explicit mappings for each value, a clear warning message if an unknown value is encountered, and then also use more natural values for the API, i.e. start with `BIT(0)`.

Reply:
Updated to start with `BIT(0)`